src/RewardsDistributor/MorphoCompoundRewardsDistributor.ts

```diff
@@ -0,0 +1,305 @@
+              accruedMorpho: BigNumber.from(userConfig.accruedMorpho),
+            },
+          ])
+        ),
+      ])
+    );
+    return distributor;
+  }
+
+  /**
+   * Init function fetch the events before the first epoch in order
+   * to retrieve initial user states
+   */
+  static async init(provider: providers.Provider, morphoCompound: MorphoCompound) {
+    const firstEpochBlock = allEpochs[0].initialBlock!;
+
+    const distributor = new MorphoCompoundRewardsDistributor(provider, morphoCompound);
+    // Initialize the balances at the beginning of the first epoch
+    await distributor.batchEvents(MORPHO_COMPOUND_DEPLOYMENT_BLOCK, firstEpochBlock, false);
+    return distributor;
+  }
+
+  // ----------------- storage ------------------------
+  public BATCH_SIZE = 100_000; // number of blocks per batch
+  private blockFetcher: BlockFetcher;
+
+  constructor(private provider: providers.Provider, private morphoCompound: MorphoCompound) {
+    super(MORPHO_COMPOUND_DEPLOYMENT_BLOCK);
+    this.blockFetcher = new BlockFetcher(provider);
+  }
+
+  async applyEpoch(epoch: EpochConfig) {
+    if (!epoch.initialBlock) throw Error(`No initial block for epoch ${epoch.id}`);
+    if (!epoch.finalBlock) throw Error(`No final block for epoch ${epoch.id}`);
+    if (this.lastBlockSynced >= epoch.finalBlock) {
+      // epoch already applied
+      return;
+    }
+    this.currentEpoch = epoch;
+    return this.batchEvents(epoch.initialBlock, epoch.finalBlock, true);
+  }
+
+  async batchEvents(blockFrom: number, blockTo: number, updateRewards = true) {
+    if (blockTo > blockFrom + this.BATCH_SIZE) {
+      let block = blockFrom;
+      while (block < blockTo) {
+        await this.batchEvents(block, Math.min(block + this.BATCH_SIZE, blockTo));
+        block += this.BATCH_SIZE;
+      }
+      return;
+    }
+    // process sequentially to prevent throughput issues
+    const allEvents = await Promise.all([
+      await this.morphoCompound.queryFilter(this.morphoCompound.filters.Repaid(), blockFrom, blockTo),
+      this.morphoCompound.queryFilter(this.morphoCompound.filters.Borrowed(), blockFrom, blockTo),
+      this.morphoCompound.queryFilter(this.morphoCompound.filters.Supplied(), blockFrom, blockTo),
+      this.morphoCompound.queryFilter(this.morphoCompound.filters.Withdrawn(), blockFrom, blockTo),
+      this.morphoCompound.queryFilter(this.morphoCompound.filters.P2PIndexesUpdated(), blockFrom, blockTo),
+    ]);
+
+    const allEventsSorted = allEvents.flat().sort((eventA, eventB) => {
+      if (eventA.blockNumber !== eventB.blockNumber) return eventA.blockNumber > eventB.blockNumber ? 1 : -1;
+      // events in the same block
+      if (eventA.transactionIndex !== eventB.transactionIndex)
+        return eventA.transactionIndex > eventB.transactionIndex ? 1 : -1;
+      // events in the same tx, ordered by log index
+      return eventA.logIndex > eventB.logIndex ? 1 : -1;
+    });
```

Comment:
```suggestion
    const allEventsSorted = _sortBy(allEvents.flat(), ['blockNumber', 'transactionIndex', 'logIndex'])
```

see https://codesandbox.io/s/lodash-playground-forked-kthyu4?file=/src/index.js